### ISSUE TITLE
Bug in the type checker

### ISSUE THREAD

**Issue by dasch** [elm-lang/elm-compiler#1204]
I was getting a weird type error when trying to compile this code:

```elm
module Stream where


import Dict exposing (Dict)


type alias Stream comparable v = Signal (comparable, v)


leftJoin : Stream comparable v1 -> Stream comparable v2 -> Stream comparable (v1, Maybe v2)
leftJoin stream1 stream2 =
  let
      buffer : Signal (Dict comparable v2)
      buffer = Signal.foldp step Dict.empty stream2

      step : (comparable, v2) -> Dict comparable v2 -> Dict comparable v2
      step (k, v2) = Dict.insert k v2

      lookup : (comparable, v1) -> Dict comparable v2 -> (comparable, (v1, Maybe v2))
      lookup (k, v1) state = (k, (v1, Dict.get k state))
  in
      Signal.map2 lookup stream1 buffer
```

The error was:

```
The type annotation for `leftJoin` does not match its definition.

10│ leftJoin : Stream comparable v1 -> Stream comparable v2 -> Stream comparable (v1, Maybe v2)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The type annotation is saying:

    Signal ( comparable, a )
    -> Signal ( comparable, b )
    -> Signal ( comparable, ( a, Maybe b ) )

But I am inferring that the definition has this type:

    Signal ( comparable, a )
    -> Signal ( comparable, a )
    -> Signal ( comparable, ( a, Maybe b ) )
```

Changing `buffer` to this and calling it with `stream2` fixes the problem:

```elm
buffer : Stream comparable v2 -> Signal (Dict comparable v2)
buffer stream = Signal.foldp step Dict.empty stream
```

That doesn't seem very intuitive to me – seems to be a compiler bug.

**Comment by jvoigtlaender** [comment]
What happens if you omit the type signatures for the local functions `buffer`, `step`, `lookup`?

Hint: `buffer : Signal (Dict comparable v2)` there, for example, does not mean what you probably think it means. It does not mean that the `v2` will be the one from the signature of `leftJoin`. Instead, your program is equivalent to one where you annotate the local function like so: `buffer : Signal (Dict comparable some_completely_new_type_variable)`. Would you be less surprised, than you have been, that *that* program is not accepted by the compiler?

The reason for all this is that Elm has no equivalent to Haskell's https://wiki.haskell.org/Scoped_type_variables.

In conclusion, I doubt you have found a compiler bug here. You have come upon a need for a feature that the compiler does not have.

**Comment by dasch** [comment]
I that the type variables are not bound just because they're in a let – but that wasn't the problem. I had this:

```elm
f x =
  let
    y = something x
  in
    somethingElse y
```

... which I had to change to

```elm
f x =
  let
    y x = something x
  in
    somethingElse (y x)
```

**Comment by rgrempel** [comment]
I just tried this in elm-lan.org/try for fun, and it seems to work as expected:

```elm
import Graphics.Element exposing (show)

f x =
  let
    y = something x
  in
    somethingElse y
    
something x =
  "something" ++ x
   
somethingElse y =
  "somethingElse" ++ y
  
main = show (f "base")
```

Then, for even more fun, I tried your original example, deleting the type annotations for the functions inside the `let`, as Janis suggested. It compiled without error.

**Comment by rgrempel** [comment]
BTW, at least in Elm 0.16, the compiler error for your original example ends with this text:

```
Hint: A type annotation is clashing with itself or with a sub-annotation. This
can be particularly tricky, so read more about it.
<https://github.com/elm-lang/elm-compiler/blob/0.16.0/hints/type-annotations.md>
```

And, if you go to the referenced page, the third hint points to the problem that Janis mentioned.

So, more kudos for the Elm 0.16 error messages!

**Comment by evancz** [comment]
Yeah, this is a known thing. I opened #1214 to track the root issue here.

If you want, it may make sense to browse through https://github.com/elm-lang/error-message-catalog/issues to see if anyone else has reported this. If not, it'd be good to open something similar with an http://sscce.org so people can find their way to the root thing.
